### PR TITLE
Refactor Istiocoredns configuration and support other FQDN (multicluster)

### PR DESCRIPTION
**Pull request by zufardhiyaulhaq**
Issue: https://github.com/istio/istio/issues/25324

Refactor Istiocoredns configuration to support other FQDN.

[X] Configuration Infrastructure
[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by shamsher31** [comment]
@zufardhiyaulhaq  Please resolve the conflict.

**Comment by zufardhiyaulhaq** [comment]
@shamsher31 please check, thank you.

File: manifests/charts/istiocoredns/values.yaml
Comment (shamsher31) [inline manifests/charts/istiocoredns/values.yaml:47]: Can you sort this alphabetically?

**Review by shamsher31 — CHANGES_REQUESTED** [review]

File: manifests/charts/istiocoredns/values.yaml
Comment (zufardhiyaulhaq) [reply manifests/charts/istiocoredns/values.yaml:47]: sure, let me do that

Comment (shamsher31) [reply manifests/charts/istiocoredns/values.yaml:47]: It would be good to add some description for the field and what it does eg: [like this](https://github.com/istio/istio/blob/master/manifests/examples/multicluster/values-istio-multicluster-gateways.yaml#L16).

Comment (zufardhiyaulhaq) [reply manifests/charts/istiocoredns/values.yaml]: done, waiting for CI

Comment (zufardhiyaulhaq) [reply manifests/charts/istiocoredns/values.yaml]: done. please check

**Comment by zufardhiyaulhaq** [comment]
@shamsher31 please check. thank you

**Comment by shamsher31** [comment]
I will let @istio/wg-environments-maintainers to take a look. Thanks

File: manifests/charts/istiocoredns/templates/configmap.yaml
Comment (sdake) [inline manifests/charts/istiocoredns/templates/configmap.yaml:16]: docs state this is a space separate list, lgtm.

File: manifests/charts/istiocoredns/templates/configmap.yaml
Comment (sdake) [inline manifests/charts/istiocoredns/templates/configmap.yaml:24]: I don't understand this logic. The prior logic was conditional on the tag in use. This is not conditional on the tag in use. Can you expand on what your goal with this snippet is?

File: manifests/charts/istiocoredns/values.yaml
Comment (sdake) [inline manifests/charts/istiocoredns/values.yaml:110]: do you ever expect this list to expand? There is a general problem with this approach in that K8s has a limit of 6 search locations in /etc/resolv.conf. This makes expanding this list difficult, and in general even using this list difficult. The reason is that many cloud providers already use 4 to 5 search spaces, so adding 1, while not always bad, may not always work. Adding 2 or more can crater the entire deployment.

**Review by sdake — COMMENTED** [review]
some concerns about the amount of PodDNSSearchNamespaces. The concept looks good, although in practice, K8s does not cooperate.

File: manifests/charts/istiocoredns/templates/configmap.yaml
Comment (rshriram) [reply manifests/charts/istiocoredns/templates/configmap.yaml:16]: is this even right? doesn't the dns search namespace include the kubernetes namespace where coredns is running?

Comment (rshriram) [reply manifests/charts/istiocoredns/templates/configmap.yaml:16]: oh ignore..

**Comment by rshriram** [comment]
This seems fine (though the list of search namespaces seems overkill if all you want is a replacement for .global). FYI, with #24919 you may not need core dns at all.

File: manifests/charts/istiocoredns/templates/configmap.yaml
Comment (zufardhiyaulhaq) [reply manifests/charts/istiocoredns/templates/configmap.yaml:16]: yes that will ignore the custom FQDN

File: manifests/charts/istiocoredns/templates/configmap.yaml
Comment (zufardhiyaulhaq) [reply manifests/charts/istiocoredns/templates/configmap.yaml:24]: this will create multiple server blocks for each of the FQDN. I think the istiocoredns only used by multicluster cases.

File: manifests/charts/istiocoredns/values.yaml
Comment (zufardhiyaulhaq) [reply manifests/charts/istiocoredns/values.yaml:110]: yes I expect the list to expand. what do you mean the limitation of resolv.conf. we can patch the coredns/kubedns configmap in kube-system to proxy to istiocoredns

**Review by zufardhiyaulhaq — COMMENTED** [review]
> This seems fine (though the list of search namespaces seems overkill if all you want is a replacement for .global). FYI, with #24919 you may not need core dns at all.

in which version/milestone will this apply?

**Comment by rshriram** [comment]
@zufardhiyaulhaq it will be in Istio 1.7

**Comment by zufardhiyaulhaq** [comment]
@rshriram did istiocoredns will be deprecated in 1.7? or we will still support istiocoredns?

**Review by sdake — APPROVED** [review]

File: manifests/charts/istiocoredns/values.yaml
Comment (sdake) [reply manifests/charts/istiocoredns/values.yaml:110]: https://github.com/istio/istio/blob/master/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml#L416-L421 will cause every sidecar's resolv.conf to be injected with DNSConfig's list. I know - because I wrote the code...

https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config

K8s "allows for at most 6 search domains". Unfortunately, I would later find out this is baked into glibc itself.

Hence, why there is no variable list, but only `.global`. There was, at one time, a variable list, but it broke most of the time on most of the clouds.

Cheers,
-steve

Comment (zufardhiyaulhaq) [reply manifests/charts/istiocoredns/values.yaml:110]: @sdake Okay so the DNSSearchNamespaces is injected to the sidecar. but how it will affect the multicluster communication? since for my understanding, the app pod make a call directly to coredns/kubedns in kube-system to resolve the domain.

**Comment by zufardhiyaulhaq** [comment]
@rshriram what do you think? did this PR is still relevant?

**Pull request closed without merging**
